paho/pinger.go

```diff
@@ -81,106 +73,67 @@ func (p *DefaultPinger) Run(conn net.Conn, keepAlive uint16) error {
 		return fmt.Errorf("conn is nil")
 	}
 	p.mu.Lock()
-	if p.timer != nil {
+	if p.running {
 		p.mu.Unlock()
-		return fmt.Errorf("Run() already called")
+		return fmt.Errorf("Run() already in progress")
 	}
-	select {
-	case <-p.done:
+	p.running = true
+	p.mu.Unlock()
+	defer func() {
+		p.mu.Lock()
+		p.running = false
 		p.mu.Unlock()
-		return fmt.Errorf("Run() called after stop()")
-	default:
+	}()
+
+	interval := time.Duration(keepAlive) * time.Second
+	timer := time.NewTimer(0) // Immediately send first pingreq
+	var lastPingSent time.Time
+	for {
+		select {
+		case <-ctx.Done():
+			timer.Stop() // We don't care if the timer has fired
+			return nil
+		case t := <-timer.C:
+			p.mu.Lock()
+			lastPingResponse := p.lastPingResponse
+			pingDue := p.lastPacketSent.Add(interval)
+			p.mu.Unlock()
+
+			if !lastPingSent.IsZero() && lastPingSent.After(lastPingResponse) {
+				p.debug.Printf("DefaultPinger PINGRESP timeout")
+				return fmt.Errorf("PINGRESP timed out")
+			}
+
+			if t.Before(pingDue) {
+				// A Control Packet has been sent since we last checked, meaning the ping can be delayed
+				timer.Reset(pingDue.Sub(t))
+				continue
+			}
+
+			if _, err := packets.NewControlPacket(packets.PINGREQ).WriteTo(conn); err != nil {
```

Comment:
can this potentially block for a very long time, breaking the keep alive mechanism?

Reply:
If the connection is mostly idle, we rely on keepalives to detect network issues. PingHandler sends a PINGREQ, and if it doesn't receive a response within KeepAlive period, the client will be closed. However, this logic breaks in the edge cases when PingHandler hangs on the WriteTo(conn) operation due to some network issues; the call will be eventually unblocked due to the TCP timeout, but it could be way longer than the specified KeppAlive period. 

Isn't it better to send a packet in a separate goroutine; and in case a response is not received within the KeepAlive period, close the connection (which would unblock the potentially stuck goroutine)?

Reply:
>Isn't it better to send a packet in a separate goroutine

OK - when you said it could block I thought you meant it was blocked because a message was being transmitted. This can happen, if you send a large message over a slow link there is a potential for it to take longer than  `keepAlive` to be transmitted (and there is not much we can do about that).

I believe the defaults for TCP retransmission periods are something like 13 to 30 minutes on Linux and around 65 seconds on windows. so there does appear to be a potential for the  `Write` to block longer than the keepalive period.

So there may be a rationale for moving the `Send` to a goroutine but I think that should be a separate issue/PR (this PR was more about tidying things up and removing a potential deadlock so the way the packet is sent was not changed from previous code).

I would note that the much older `paho.mqtt.golang` sends pings in the same way (does not include any checks re how long the send takes) and I don't believe any issues have been raised.

Reply:
Created a [PR](https://github.com/eclipse/paho.golang/pull/257).